Add blackbox tests and documentation

acoulton opened the pull request:
* [x] get the blackbox correctly running with lowest version of dependencies in CI (or skip the --lowest build to avoid confusion)
* [x] basic documentation of how the blackbox tests are wired
* [x] test configuring timezone and locale
* [x] ideally test logging (at least basic)

acoulton committed:
[WIP] Add initial blackbox test

acoulton committed:
Refactor commonality from initial smoketest

acoulton committed:
Add blackbox coverage of HTTP responses & headers

acoulton committed:
Add blackbox coverage of parsing incoming requests

acoulton committed:
drop pending blackbox

acoulton committed:
Add tests for error handling and output buffering

acoulton committed:
Throw instead of silently overwriting if attempting to send duplicate headers

We can't reliably guess whether or not to pass replace:true when setting
headers before sending the response. Previously this was just overwriting
and sending the last value provided by the handler. Instead, throw an
exception to be explicit that this is not supported.

acoulton committed:
Improve handling if code starts output before the framework even boots

Not much we can do here, the headers are already sent so we can't make it
a 500 but we should at least make sure the logs and any output reflects the
actual error that happened rather than causing a second error trying to
send headers again.

acoulton committed:
Name the dynamic blackbox handlers consistently

Rather than giving them random names for every run, provision them
based on the testcase and hash of the content, so that it's easier
to find them locally for debugging and we can avoid having to drop
and recreate them all the time.

acoulton committed:
Fix cleaning up after an interactive blackbox run

acoulton committed:
Add blackbox coverage of logging

Cover the key functionality of the stackdriver logger, including request
logging, exception logging, and custom logging. I've tested with this
because it's the only logger microframework directly supports, and because
it's complex enough that we should be able to have confidence any generic
PSR\LoggerInterface would work as expected if wired up correctly.

acoulton committed:
Add blackbox coverage of the framework setting timezone

acoulton committed:
Add basic documentation of the blackbox tests

acoulton committed:
Wait a moment for logs before asserting them

The blackbox test was flickering as logs seem to sometimes take
a few moments to arrive on disk. Put in a short sleep
to give them a chance to appear before we start any
assertions on their content.

acoulton commented:
The only thing I still haven't resolved is how to neatly run the blackbox with the `--lowest` dependency versions. I can probably just pass the environment variable along through the docker build but it'll be a bit ugly.

I'm considering whether for this project it's better to either:

* just remove the `--lowest` build outright
* keep the `--lowest` build but skip the blackbox tests step (so it's at least obvious if you look at the actions output that it didn't run)

craig410 commented on src/MicroFramework.php:
```diff
@@ -162,8 +165,11 @@ private function logUncaughtException(LoggerInterface $logger, Throwable $e): vo
     {
-        http_response_code(500);
-        header('Content-Type: text/plain');
+        if (!headers_sent()) {
```
Same whitespace `!`. ...Actually there is quite a lot throughout the pull

acoulton replied:
Urgh IDE settings didn't carry over when I opened as a new project :-( will fix

craig410 commented on src/MicroFramework.php:
```diff
@@ -111,7 +112,7 @@ private function resetOutputBuffersAssertingNoOutput(): void
         while (ob_get_level()) {
             $buffer = ob_get_clean();
-            if ( ! empty($buffer)) {
+            if (!empty($buffer)) {
```
The whitespace changes have lost the space around `!` which makes it harder to spot / isn't our usual style

craig410 commented on composer.json:
```diff
@@ -10,7 +10,9 @@
     "fig/log-test": "^1.1",
-    "phpunit/phpunit": "^9.6"
+    "guzzlehttp/guzzle": "^7.8",
+    "phpunit/phpunit": "^9.6",
```
Should probably start using `phpunit:^10.0` here

craig410 commented on test/blackbox/docker-compose.yaml:
```diff
@@ -0,0 +1,103 @@
+  # The test_runner is the outside of the black box, it has a mounted copy of the full
+  # codebase for running tests etc.
+  test_runner:
+    image: php:8.2-alpine
```
Don't know if this image is already cached locally in github actions but given you need `php:8.2-apache` - which is based on debian - in the next container it would save pulling an another image and an additional 40Mb here and in dev.

craig410 commented on test/blackbox/docker-compose.yaml:
```diff
@@ -0,0 +1,103 @@
+
+  test_subject_log_receiver:
+    # No real need to use the php image - it's just running netcat as a simple TCP > file dump - but saves another pull
+    image: php:8.2-alpine
```
could just use `alpine:latest` it's layers will already be there unless you take on my comment above but it's only 3Mb if you want to be explicit it's got nothing to do with PHP

craig410 commented on test/blackbox/logging/log_receiver.sh:
```diff
@@ -0,0 +1,8 @@
+#!/bin/ash
```
is this a typo or do you really want to use `ash`?  It's probably just a symlink to `sh` anyway on most systems these days

acoulton replied:
Just that I've always tended to use /bin/bash by default and similarly then started using ash on alpine. Will change to sh - don't think I'm using any shell-specific features

craig410 replied:
Oh I don't think I'd twigged you were on alpine / busybox in here. Presumed it was a debian container with bash installed. Leave it as is.

craig410 commented on test/run-blackbox-tests-interactive.sh:
```diff
@@ -0,0 +1,27 @@
+echo "Starting interactive blackbox test runner shell"
+echo "Hint: Run docker compose logs -f in another terminal to see what's happening"
+echo "Hint: Changes to tests are mounted live, but changes to application code *will not* apply"
+echo "      until you re-build the containers (e.g. by exiting and re-running this script)"
```
Would using https://docs.docker.com/compose/file-watch/ help with not having to manually quit, teardown, rebuild and run?

acoulton replied:
Not as-is, because you also need to commit changes to the `src` to have `git archive` include them. It would be possible to have the docker build do a `git stash` and then archive the stash rather than HEAD but would only want to do that in dev so it means more flags and complexity to the build process... Will clarify why not in the docs.

craig410 commented on test/blackbox/package-for-blackbox-tests.sh:
```diff
@@ -0,0 +1,33 @@
+# version of the library and package it up with the other dependencies ready for running HTTP
+# tests. It runs in a *composer* image, not our own, and is used as the first part of a
+# multi-stage build. See the docker-compose.yaml for how this fits together.
+
```
I would extract or duplicate the bones of this comment and the ones below up to the README as I was pretty confused about where the `git export` was actually running when I was reading it until I found these comments.

acoulton replied:
Yeah was planning to add an overall readme and point to this.

craig410 commented on test/blackbox/README.md:
```diff
@@ -0,0 +1,28 @@
+Note that this means when working locally, changes to the application code **will not be picked up by the test_subject 
+automatically**. To update the test_subject you will need to **commit your changes** (so that they are included in the
+`git export HEAD`) and then **exit and re-provision the blackbox testing stack**.
+
```
Does using `git export` here prevent us using `.gitattributes` to exclude the `/test` and `/.github` dirs from the package dist, so we have to carry all this extra weight around everywhere?

Alternatively - I guess we could exclude these dirs in `.dockerignore` in the projects using this package to achieve a similar effect if needs be

I would also add something to this comment about _why_ we're using git export - is it purely because we have to bring it back in through composer or because we're explicitly trying to test the built package.

acoulton replied:
First off there's a typo/brainfart in this readme that it's `git archive` not `git export`. And I'll clarify that the point of using it is because that then produces the exact same tar that will be produced by GitHub and consumed by composer - e.g. respecting the .gitattrributes and ignored files - so is the key step that allows us to blackbox test the distribution version of the package rather than the files we have checked out locally.

craig410 replied:
Ahh ok, it's all starting to make a lot more sense now.

craig410 replied:
We should probably add a .gitattributes at this point too so we can see what is excluded

acoulton replied:
There's one already, from the initial version https://github.com/ingenerator/microframework/blob/1.x/.gitattributes - `/test` is excluded from the export.

I should add the .github dir too

craig410 commented:
> The only thing I still haven't resolved is how to neatly run the blackbox with the `--lowest` dependency versions. I can probably just pass the environment variable along through the docker build but it'll be a bit ugly.
> 
> I'm considering whether for this project it's better to either:
> 
> * just remove the `--lowest` build outright
> * keep the `--lowest` build but skip the blackbox tests step (so it's at least obvious if you look at the actions output that it didn't run)

Either of those options would be better than trying to run the blackbox tests with another flag. There aren't many dependencies of this for it to go too wrong either way.

acoulton committed:
Only build with current versions of dependencies

The `--lowest` build won't apply as-is to the blackbox tests anyway,
and would require a bit of fiddly wiring to carry that config over.
For this package, the dependencies are limited and will be kept
fairly tight, so there's limited value in the extra complexity of
running a separate --lowest build.

acoulton committed:
Exclude the .github from exported package

acoulton committed:
Add .editorconfig to maintain code style

acoulton committed:
Fix code style of new code to match editorconfig

acoulton committed:
Simplify number of docker images used for blackbox

Use the apache image as the test runner, since we need
to pull it anyway.

acoulton committed:
Update to phpunit 10

acoulton committed:
Update blackbox testing documentation

acoulton committed:
Add main README

acoulton committed:
Bump to latest supported versions

We intend to keep the dependency range for this package very tight.

craig410 reviewed (approved):
LGTM. Don't know if you want to bump the changelog date or not?

acoulton committed:
Bump changelog
